### PR TITLE
Experimental NLP metric

### PR DESCRIPTION
**Pull request by richard-rogers**
## Description

Adds
* `VectorComponent` for holding matrices and vectors
* `SvdMetricConfig`, `SvdMetric`, and `UpdatableSvdMetric` for vector-space embeddings
* `NlpConfig`, `BagOfWordsMetric`, `LsiMetric` for NLP logging
* Adds `NumpyView.strings` to support bag of words logging
* Moves `EmbeddingMetric` to `whylogs/experimental/extras`

### CODE REVIEW

File: python/examples/experimental/NLP_Summarization.ipynb [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "yw9ErM4ZTpZo"
+      },
+      "outputs": [],
+      "source": [
+        "%pip install nltk\n",
+        "%pip install bs4\n",
+        "%pip install whylogs-1.1.19-py3-none-any.whl "
```
Comment (richard-rogers) [inline python/examples/experimental/NLP_Summarization.ipynb:66]: Update when wheel is released

File: python/examples/experimental/NLP_Summarization.ipynb [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+      "metadata": {
+        "id": "TrF-XmKsn35M"
+      },
+      "execution_count": 4,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "We also need to attach the `BagOfWordsMetric` to the columns that represent our input articles and output summaries. Unfortunately, whylogs doesn't fully understand bags of words as loggable values yet, so we'll also bump up the Python logging level to suppress a spurious warning message."
```
Comment (richard-rogers) [inline python/examples/experimental/NLP_Summarization.ipynb:287]: This should be better post-embeddings work.

File: python/whylogs/experimental/extras/nlp_metric.py [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+    # should also zap if too small relative to s[0]
+    return np.array([1 / x if x > _SMALL else 0 for x in s])
+
+
+def _serialize_ndarray(a: np.ndarray) -> bytes:
+    bio = BytesIO()
+    np.save(bio, a, allow_pickle=False)
+    return bio.getvalue()
+
+
```
Comment (richard-rogers) [inline python/whylogs/experimental/extras/nlp_metric.py]: These should share the `MatrixComponent` stuff from the embeddings work.

File: python/whylogs/experimental/extras/nlp_metric.py [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+        # TODO: make both metrics optional?
+        if name.endswith("_bag_of_words"):
+            return {BagOfWordsMetric.get_namespace(): BagOfWordsMetric.zero(column_schema.cfg)}
+        elif name.endswith("_lsi"):
+            return {LsiMetric.get_namespace(): LsiMetric.zero(column_schema.cfg)}
+        return self._resolver.resolve(name, why_type, column_schema)
+
+
+class NlpLogger:
+    def __init__(
```
Comment (richard-rogers) [inline python/whylogs/experimental/extras/nlp_metric.py:353]: Maybe we can drop this now?

File: python/examples/experimental/NLP_Summarization.ipynb (hunk 1 above)
Comment (jamie256) [reply python/examples/experimental/NLP_Summarization.ipynb:66]: you could release a dev version that unblocks this? or put in a version range:
`%pip install "whylogs>1.1.21" --pre`

**Review by jamie256 — DISMISSED** [review]
Looks good for experimental namespace. We should follow up with an extra for the dependencies used here, but we can address in a follow up change.

File: python/whylogs/core/preprocessing.py [hunk 5]
```diff
@@ -34,10 +34,11 @@ def iterables(self) -> List[List[Any]]:
 class NumpyView:
     ints: Optional[np.ndarray] = None
     floats: Optional[np.ndarray] = None
+    strings: Optional[np.array] = None
```
Comment (richard-rogers) [inline python/whylogs/core/preprocessing.py]:
```suggestion
    strings: Optional[np.ndarray] = None
```

File: python/whylogs/core/preprocessing.py [hunk 6]
```diff
@@ -239,11 +242,13 @@ def apply(data: Any) -> "PreprocessedColumn":
 
         if isinstance(data, np.ndarray):
             result.len = len(data)
-            if issubclass(data.dtype.type, np.number):
+            if issubclass(data.dtype.type, (np.number, np.str_)):
```
Comment (richard-rogers) [inline python/whylogs/core/preprocessing.py:245]: I'm not sure this is correct.

File: python/whylogs/experimental/extras/nlp_metric.py [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+        U = np.dot(Q, UR)
+        return U, S
+
+    def merge(self, other: "SvdMetric") -> "UpdatableSvdMetric":
+        # other can be updatable or not
+        new_U, new_S = self._resketch(self.k.value, self.decay.value, other.U.value, other.S.value)
+        return UpdatableSvdMetric(self.k, self.decay, MatrixComponent(new_U), MatrixComponent(new_S))
+
+    def columnar_update(self, data: PreprocessedColumn) -> OperationResult:
+        if not data.list.objs:
```
Comment (richard-rogers) [inline python/whylogs/experimental/extras/nlp_metric.py:143]: use `NumpyView`

**Review by jamie256 — APPROVED** [review]
Still looks ok after rebase and updates.

**Pull request merged**